# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [4]:
def addtwo(x):
    return x+2
print(addtwo(5))

#the lambda version....
#without using a variable (g= somthing) to pass in data

print((lambda x: x+2)(5)) #lambda sets x ==5 and runs the function

#using a variable

num=4
print((lambda x: x+3)(num))

#### Lambda basicaly says "put your variable here! its a reciever for a function"

7
7
7


#### Saving to a Variable

In [6]:
f_test = lambda x: x+2

f_test(4)

6

#### Multiple Inputs

In [15]:
#multiple inputs with no variable
print((lambda x, y, z: x * y * z)(6, 8 , 77))

#multiple inputs with a variable attatched
x=lambda x, y, z: x * y * z
print(x(3,5,8))

3696
120


#### Passing a Lambda into a Function

In [13]:
def multiply(f, num):
    return f(num)
def subtraction(t,num):
    return t(num)
    """
        f expects a lmbda funtion
        num expects a number
    """
multiply(lambda x: x*x, 4)
multiply(lambda x: x*x,7)
subtraction(lambda x: 8-x,5)

3

#### Returning a Lambda from a Function

In [21]:
# regular defined function
def multiply_test(num):
    return num * 4

#function within a function:
def returnfunc():
    test = 4
    def multiply(num):
        return num *2
    return multiply
f_return=returnfunc()

print(returnfunc())
print(f_return(5)) #this is a lot of writing and is unesseccary

#lambda function returned from a regular function
def returnLam(b,c):
    return lambda x, a: x+a+b+c #we put a lambda function within a function
r_lamb = returnLam(4,5) #we preset b and c as they are constants

print(r_lamb(5,5)) #all thats left is b and c

<function returnfunc.<locals>.multiply at 0x0000026977E35760>
10
19


#### If Statements within Lambdas

In [24]:
# lambda x: True if (condition) else false
f_condition = lambda num : num * 2 if num >10 else num +2

print(f_condition(9))

11


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [30]:
f= lambda x: x**3
f(3)

27

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [33]:
#map(function,iterable)
#an iterable is a list, dict, tuple, set. etc
#a map is usually used with a pre-defined function, but we can use lambdas as well
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
        
numbers=[6,4,90,77,46,22,20,4]
more_nums=[8,3,1,55,3,63,24,53]

squared_map = map(squared,numbers,more_nums)
print(squared_map)
squared_map1=list(map(squared, numbers, more_nums)) #it takes one number from each list at a time and runs them through
                                                    #the function in sets
print(squared_map1)

[(36, 64), (16, 9), (90, 1), (77, 55), (46, 3), (22, 63), (20, 24), (4, 53)]


#### Using Lambda's with Map

In [36]:
#map(lambda x: do sonmthing, iterable)
#using a map with lambda happens inline or or one line
#list(map(lambda x: x*2, nums))
list_nums_lambda=list(map(lambda x ,y: (x**2, y**2) if x<10 and y<10 else (x,y), numbers, more_nums))
print(list_nums_lambda)

[(36, 64), (16, 9), (90, 1), (77, 55), (46, 3), (22, 63), (20, 24), (4, 53)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [3]:
num=[1,2,3,4,5,6,7,8]

double_minus=list(map(lambda x: (x*2)-1,num))
print(double_minus)

#or

double_minus=list(map(lambda x: (x*2 , x-1),num))
print(double_minus)

[1, 3, 5, 7, 9, 11, 13, 15]
[(2, 0), (4, 1), (6, 2), (8, 3), (10, 4), (12, 5), (14, 6), (16, 7)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [47]:
names=["Andrew","Kenny","Nick","Chase","Eric","Kingsley"]

def a_names(name):
    if name[0].lower()== "e":
        return True
    else:
        return False
    
new_names_list=list(filter(a_names,names))
print(new_names_list)

['Eric']


#### Using Lambda's with Filter()

In [48]:
new_names_lambda = list(filter(lambda name: True if name[0].lower()=='a' else False, names))
print(new_names_lambda)

['Andrew']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [60]:
import statistics
numbs=[1,2,3,4,5,6]
mean= sum(numbs)/6
print(mean)
answer=list(filter(lambda x:x >= mean,numbs))
print(answer)

3.5
[4, 5, 6]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [7]:
from functools import reduce
#reduce(function,iterable)

numbers=[1,2,3,4,5,6,4]

def sum(num1, num2):
    return num1 + num2

answer=reduce(sum,numbers) #this is basically a for loop that runs items in a list through a machine.
#in this case it runs every number throught the function (sum) and adds everything together.
print(answer)

def sub(num1,num2):
    return num1- num2
new_answer= reduce(sub, numbers)
print(new_answer)

25
-23


#### Using Lambda's with Reduce()

In [9]:
subtract= reduce(lambda x , y: x-y,numbers)
print(subtract)

-23


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [10]:
my_list=[1, 2, 3, 4]

multi= reduce(lambda x, y: x*y,my_list)
print(multi)


24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [9]:
def addnums(number):
    #here is your base case
    if number <= 1:
        print('addnums(1) = 1')
        return number
    else: 
        print( f"addnums({number})={number} + addnums({number - 1})")
        return number + addnums(number - 1)
    
addnums(5)

addnums(5)=5 + addnums(4)
addnums(4)=4 + addnums(3)
addnums(3)=3 + addnums(2)
addnums(2)=2 + addnums(1)
addnums(1) = 1


15

#### Writing a Factorial Function

In [10]:
# 5! means 5 factored (5 * 4 * 3 * 2 * 1)
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)#this is abasically a loop that says, "untill we reach the numer 1, keep 
                                        #keep going" and it does this by calling on itself.
        
factorial(7)

5040

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [32]:
def reduction(number):
    if number == 1:
        print("all done")
        return number
    else:
        print(f'reduction({number})={number} - reduction({number-1})')
        return number - reduction(number-1)

    
reduction(50)

reduction(50)=50 - reduction(49)
reduction(49)=49 - reduction(48)
reduction(48)=48 - reduction(47)
reduction(47)=47 - reduction(46)
reduction(46)=46 - reduction(45)
reduction(45)=45 - reduction(44)
reduction(44)=44 - reduction(43)
reduction(43)=43 - reduction(42)
reduction(42)=42 - reduction(41)
reduction(41)=41 - reduction(40)
reduction(40)=40 - reduction(39)
reduction(39)=39 - reduction(38)
reduction(38)=38 - reduction(37)
reduction(37)=37 - reduction(36)
reduction(36)=36 - reduction(35)
reduction(35)=35 - reduction(34)
reduction(34)=34 - reduction(33)
reduction(33)=33 - reduction(32)
reduction(32)=32 - reduction(31)
reduction(31)=31 - reduction(30)
reduction(30)=30 - reduction(29)
reduction(29)=29 - reduction(28)
reduction(28)=28 - reduction(27)
reduction(27)=27 - reduction(26)
reduction(26)=26 - reduction(25)
reduction(25)=25 - reduction(24)
reduction(24)=24 - reduction(23)
reduction(23)=23 - reduction(22)
reduction(22)=22 - reduction(21)
reduction(21)=21 - reduction(20)
reduction(

25

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [21]:
def my_range(stop, start, step=3):
    while start<stop:
        yield start #this word yield denotes or initializes a generator
        start += step
for i in my_range(36,1):#the step number is defaulted in the function already
    my_generator_value= i
    print(my_generator_value)
###!!!!! generators allow you to create functions that create iterators 

1
4
7
10
13
16
19
22
25
28
31
34


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [23]:
def squared(stop,start,step=1):
    while start > stop:
        yield start
        start -= 1
        
for i in squared(0,5):
    my_generator_value = i**2
    
    print(my_generator_value)


25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [36]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
def nospace(space):
    if space == "  ":
        return False
    elif space == " ":
        return False
    elif space == "":
        return False
    else:
        return True
new_list=list(filter(nospace, places))
print(new_list)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [40]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]

sorted_authors = sorted(author, key = lambda name: name.split()[-1].lower())
print(sorted_authors)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [41]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# Conversion function
def celsius_to_fahrenheit(c):
    return (c * 9/5) + 32


converted_places = list(map(lambda x: (x[0], celsius_to_fahrenheit(x[1])), places))
# this line seperates the name of the town by its index, then applys the function to the thing in the second index.
print(converted_places)



[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [46]:
def fib(n):
    if n < 0:
        return
    elif n == 0:
        print("Iteration 0: 1")
        return
    elif n == 1:
        print("Iteration 0: 1")
        print("Iteration 1: 1")
        return
    else:
        sequence = [1, 1]
        for i in range(2, n + 1):
            next_value = sequence[-1] + sequence[-2]
            sequence.append(next_value)

        for i, value in enumerate(sequence):
            print(f"Iteration {i}: {value}")
fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
